src/org/rascalmpl/uri/file/MavenRepositoryURIResolver.java

```diff
@@ -138,92 +122,177 @@ private static String computeMavenCommandName() {
-                var pathIsJarRoot = "/!".equals(path);
-
-                // compensate for additional !'s produced by the previous
-                if (!pathIsJarRoot && path.startsWith("/!")) {
-                    path = path.substring(2);
-                }
+            throw new IOException("Pattern mvn:///groupId!artifactId!version did not match on " + input);
+        }
+    }
+
+    private ISourceLocation resolveInside(ISourceLocation input) throws IOException {
+        var jarLocation = resolveJar(input);
+        return URIUtil.getChildLocation(JarURIResolver.jarify(jarLocation), input.getPath());
+    }
+
+    @Override
+    public ClassLoader getClassLoader(ISourceLocation loc, ClassLoader parent) throws IOException {
+        // we simply request a classloader for the entire jar file, which will produce
+        // eventually one indexed URLClassLoader for all constituents of a classpath
+        return reg.getClassLoader(resolveJar(loc), parent);
+    }
+
+    @Override
+    public InputStream getInputStream(ISourceLocation uri) throws IOException {
+        return reg.getInputStream(resolveInside(uri));
+    }
+
+    @Override
+    public Charset getCharset(ISourceLocation uri) throws IOException {
+        return reg.getCharset(resolveInside(uri));
+    }
+
+    @Override
+    public boolean exists(ISourceLocation uri) {
+        try {
+            return reg.exists(resolveInside(uri));
+        }
+        catch (IOException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public long lastModified(ISourceLocation uri) throws IOException {
+        return reg.lastModified(resolveInside(uri));
+    }
+
+    @Override
+    public boolean isDirectory(ISourceLocation uri) {
+        try {
+            return reg.isDirectory(resolveInside(uri));
+        }
+        catch (IOException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean isFile(ISourceLocation uri) {
+        try {
+            return reg.isFile(resolveInside(uri));
+        }
+        catch (IOException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public String[] list(ISourceLocation uri) throws IOException {
+        if (uri.getAuthority().isEmpty() && (uri.getPath().isEmpty() || uri.getPath().equals("/"))) {
+            return listAllMainArtifacts();
+        }
+        else {
+            return reg.listEntries(resolveInside(uri));
+        }
+    }
+
+    /**
+     * Allows browsing through the mvn .m2 repository for "main" artifacts only
+     * @return array of all authorities encoding jar files in the .m2 repo
+     * @throws IOException if the root of the repo is gone or not permitted
+     */
+    private String[] listAllMainArtifacts() throws IOException {
+        try (Stream<Path> files = Files.walk(Paths.get(root.getPath()))) {
+            return files.filter(f -> FileNameUtils.getExtension(f).equals("jar"))
+            .filter(f -> !f.toString().endsWith("-sources"))
```

Comment:
Shouldn't it end with `sources.jar`? (Same for the line below?

Reply:
ah yes. thanks. BTW in my .m2 folder this is only 1% of the total jar files. But I'm sure there are .m2 folders with many -sources.jar files.